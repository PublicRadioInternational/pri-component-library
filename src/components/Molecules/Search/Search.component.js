/**
 * @file Search.component.js
 * Exports a search form component.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styles from './Search.css';
import ButtonInput from '../../Atoms/Button/ButtonInput.component';

/**
 * Component that renders a Search form.
 */
export default class Search extends Component {
  static propTypes = {
    baseUrl: PropTypes.string
  };

  static defaultProps = {
    baseUrl: 'https://www.pri.org'
  };

  constructor(props) {
    super(props);
    this.state = {
      searchFieldValue: null
    };
    this.handleFieldUpdate = this.handleFieldUpdate.bind(this);
    this.handleFormSubmit = this.handleFormSubmit.bind(this);
  }

  /**
   * Handles field update actions.
   *
   * @param {object} event - Event object from action.
   */
  handleFieldUpdate = event => {
    event.persist();
    this.setState(() => ({
      searchFieldValue: event.target.value || null
    }));
  };

  /**
   * Handles search form submit.
   *
   * @param {object} event - Event object from action.
   */
  handleFormSubmit = event => {
    event.preventDefault();
    const { searchFieldValue } = this.state;
    const { baseUrl } = this.props;

    // TODO: Once the search page is built out in the new frontend, this should
    // be adjusted such that it uses the correct routing methods.
    window.location.href = `${baseUrl}/search/node?search_api_views_fulltext=${searchFieldValue}`;
  };

  render() {
    return (
      <form className={styles.search} onSubmit={this.handleFormSubmit}>
        <input
          type="text"
          className={styles.searchText}
          placeholder="Search for..."
          onChange={this.handleFieldUpdate}
        />
        <ButtonInput value="Go!" className={styles.searchBtn} />
      </form>
    );
  }
}
