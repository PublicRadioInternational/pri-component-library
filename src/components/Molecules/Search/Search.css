/* import colors... */
@value colors: "../../00_global/colors.css";
@value grayLighter, gray, orange, blueDark from colors;

/* import fonts... */
@value fonts: "../../00_global/fonts.css";
@value heading from fonts;

.search {
  transition: width 500ms ease-out, background-position-x 1.25s ease-in-out;
  position: relative;

  margin: 1rem;

  background-image: linear-gradient(
    45deg,
    #777 10%,
    rgba(211, 64, 30, 0.75) 20%,
    rgba(53, 113, 156, 0.5) 50%,
    blueDark 90%
  );
  background-position-x: 0;
  background-size: 4000px;

  color: white;
  font-family: heading;
  font-size: 0.9rem;
}

.search::before {
  content: '';
  transition: bottom 200ms linear;
  position: absolute;
  top: 0;
  bottom: 1px;
  left: 0;
  right: 0;

  background-color: gray;
}

.search:focus-within,
.search.hasValue {
  background-position-x: 100%;
}

.searchForm {
  display: flex;
  position: relative;
  flex-grow: 1;
  max-width: 100%;
  background-color: blueDark;
  border-bottom: 1px solid black;
}

.searchLabel {
  transition: all 400ms ease-out;
  position: absolute;
  left: 0.5em;
  bottom: 0.25em;
  z-index: 2;

  user-select: none;

  color: #ddd;
  font-family: inherit;
  font-weight: normal;

  pointer-events: none;
}

.searchLabel::after {
  content: '...';
}

.search:focus-within .searchLabel,
.search.hasValue .searchLabel {
  transform: translate(-0.5em, -1.25em);
  opacity: 0.5;
}

.search:focus-within .searchLabel::after,
.search.hasValue .searchLabel::after {
  content: '';
}

.searchInput {
  transition: padding 200ms ease-out;
  position: relative;
  display: inline-flex;
  flex-grow: 1;
  z-index: 1;
  overflow: hidden;

  margin-top: -1em;
  margin-right: 1em;
  padding: 1em 0 0.15em 0.5em;

  background: none;

  border: none;

  color: inherit;
  font-size: 1em;
  line-height: 1.5em;

  cursor: text;
}

.searchInput::after {
  content: '\00a0';
}

.searchInput:focus,
.search.hasValue .searchInput {
  outline: none;

  margin-right: 1.5em;

  white-space: nowrap;
  overflow: hidden;
}

.searchBtn {
  transition: opacity 200ms linear, transform 200ms ease-out;
  transform-origin: 50% 50%;
  transform: rotate(90deg);
  position: absolute;
  display: flex;
  justify-content: center;
  align-items: center;
  right: 0;
  bottom: 0.15em;
  width: 1.5em;
  height: 100%;
  z-index: 1;

  padding: 0;

  border: none;
  outline: none;

  background: none;

  opacity: 0;

  pointer-events: none;
  user-select: none;

  color: inherit;

  cursor: pointer;
}

.searchIcon {
  width: 1em;
  height: 1em;
  fill: currentColor;
}

.search.hasValue .searchBtn,
.searchInput:focus + .searchBtn,
.searchBtn:focus {
  transform: rotate(0);
  opacity: 0.5;
  pointer-events: all;
}

.searchBtn:hover,
.searchBtn:focus {
  transition-delay: unset;
  opacity: 1 !important;
}

.searchBtn:active {
  color: orange;
}
