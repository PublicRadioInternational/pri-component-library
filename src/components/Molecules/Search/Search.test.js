/**
 * @file Search.test.js
 * Contains tests for Search.component.js.
 */

import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import Search from './Search.component';

describe('<Search />', () => {
  it('Handles input onChange events', () => {
    const wrapper = mount(<Search />);
    const input = wrapper.find('input[type="text"]');
    input.instance().value = 'test';
    input.simulate('change');
    expect(wrapper.state('searchFieldValue')).toBe('test');
  });

  it('Handles form submit events', () => {
    window.location.assign = jest.fn();
    const wrapper = mount(<Search />);
    const button = wrapper.find('form');
    button.simulate('submit');
    expect(window.location.assign).toHaveBeenCalledTimes(1);
  });

  it('Matches the snapshot', () => {
    const component = renderer.create(<Search baseUrl="http://google.com" />);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
