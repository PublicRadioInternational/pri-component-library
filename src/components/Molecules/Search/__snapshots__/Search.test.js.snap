// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> Matches the snapshot 1`] = `
<form
  className="search"
  onSubmit={[Function]}
>
  <label
    className="searchLabel"
    htmlFor="search"
  >
    <span
      className="visuallyhidden"
    >
      Search for...
    </span>
    <input
      className="searchText"
      id="search"
      onChange={[Function]}
      placeholder="Search for..."
      type="text"
    />
  </label>
  <div
    className="btnWrap"
  >
    <div
      className="inputWrap"
    >
      <input
        className="inputBtn btnWhite searchBtn"
        onClick={[Function]}
        type="submit"
        value="Go!"
      />
    </div>
  </div>
</form>
`;
